File: code/walk.py
# -*- coding: utf-8 
# simple random walk program

import numpy as np


class Walk:
	"""
	Random walk class at the moment in 1D and 2D only
	"""
	def __init__(self,area,n=3,thresh=1.0,M=50):
		self.Hc = thresh			#conversion parameter between walkers and concentration
		self.d = 1 if len(area)==1 else 2 	#dimension of the walk
		self.area = area	#the coordinates of the area. [[x0,x1]] in 1D, [[x0,y0],[x1,y1]] in 2D
		self.M = M 					#conversion parameter
		self.walkers = []
		self.nwalkers = len(self.walkers)
		self.first = True
		self.factor = 1e-2			#reduces the steplength 
		if len(area)==1:
			self.x0,self.x1 = area[0]
		elif len(area) == 2:
			self.x0,self.y0 = area[0]
			self.x1,self.y1 = area[1]
		# print self.X


	def HasLeftArea(self,pos):
		"""pos = [x] in 1D; 
		pos = [x,y] in 2D"""
		# print 'HasLeftArea(self,pos)'
		if self.d==1:
			if pos[0]<self.x0 or pos[0]>self.x1:
				return True
		elif self.d==2:
			if pos[0]>self.x1 or pos[0]<self.x0:
				return True
			if pos[1]>self.y1 or pos[1]<self.y0:
				return True
		return False


	def advance(self,concentration):
		"""
		will become the new advance function. 
		Concentration is now a matrix containing the entire area of the
		walk. This makes the coding simpler.
		"""
		print 'starting, %d walkers'%self.nwalkers
		if self.d !=1:
			nx,ny = np.shape(concentration)
			self.X,self.Y = np.meshgrid(np.linspace(self.x0,self.x1,nx),\
			np.linspace(self.y0,self.y1,ny)) 	#should be only if needed
		else:
			self.x = np.linspace(self.x0,self.x1,len(concentration))
		a = self.InitializeTimestep(concentration)

		steps = 100
		counter = 0
		indices = []
		walkers_leaving_area = []
		if not a:
			return np.zeros(np.shape(concentration))
		for walker in xrange(self.nwalkers):
			"loop over all walkers"	
			s = self.factor*np.random.standard_normal([self.d,steps])
			r0 = self.walkers[walker]
			for i in xrange(steps):
				# r0[:] += s[:,i]
				r0[:] = self.checkpos(r0,s[:,i])
			# for i in xrange(self.d):
			# 	r0[i] += np.sum(s[i])+self.factor*self.gradient[i]
			if self.HasLeftArea(r0):
				indices.append(counter)
				walkers_leaving_area.append(self.walkers[walker]) 
			counter += 1
		counter = 0
		# boundary = self.ReturnBoundary(walkers_leaving_area,concentration)
		boundary = self.ReturnBoundary(self.walkers,concentration)
		print 'done, %d walkers left'%self.nwalkers
		self.walkers = []
		self.nwalkers = len(self.walkers)
		boundary /= (self.M*self.Hc)
		return boundary		

	
	def InitializeTimestep(self,C):
		"""C = concentration
		should work in 1d as well now"""
		nwalkers_tmp = self.nwalkers
		C_tot = np.sum(C)
		N_walkers_tot = int(self.M*self.Hc*C_tot - nwalkers_tmp)

		if C_tot == 0 or N_walkers_tot <= 0:
			return None
		self.gradient = self.CalculateGradient(C)
		if self.d ==1:
			for i in xrange(len(C)):
				n = int((C[i]/C_tot)*N_walkers_tot)
				self.put_walkers(n,i,0)
		if self.d ==2:
			for i in xrange(len(C)):
				for j in xrange(len(C[i])):
					n = int((C[i][j]/C_tot)*N_walkers_tot)
					# print i,j
					self.put_walkers(n,i,j)
		self.nwalkers = len(self.walkers)
		return 1


	def put_walkers(self,N,i,j):
		"""Should work in 1d as well now"""
		if self.d==1:
			x = self.x[i]
			for k in xrange(N):
				self.walkers.append([x+self.factor*(0.5-np.random.uniform())])
		elif self.d==2:
			x = self.X[i,j]
			y = self.Y[i,j]
			for k in xrange(N):
				self.walkers.append([x+self.factor*(0.5-np.random.uniform()),\
					y+self.factor*(0.5-np.random.uniform())])

	def ReturnBoundary(self,walkers,concentration=None):
		"""Convert the number of walkers back into the concentration 
		at the boundary and return the boundary so it can be returned 
		from self.advance()"""
		# print concentration
		boundary = np.zeros(np.shape(concentration))
		print 'starting %d walkers'%self.nwalkers
		if self.d ==1:
			dx = self.x[1]-self.x[0]
			cont = 0
			for walker in walkers:
				cont += 1
				index = self.FindPosition(walker,dx,0)
				boundary[index] += 1
		print 'finished, have placed %d walkers'%np.sum(boundary)
		if self.d ==2:
			dx = self.X[0,1]-self.X[0,0]
			dy = self.Y[1,0]-self.Y[0,0]
			for walker in walkers:
				#find its position on the boundary
				index = self.FindPosition(walker,dx,dy)
				boundary[index[0],index[1]] += 1
		# print boundary, np.sum(boundary)
		return boundary

	def FindPosition(self,pos,dx,dy):
		"""Finds which index the walker belongs to. 
		Implements periodic boundary conditions on the walk-area
		Should work in 1d as well now"""
		indx = [-1,-1]
		if self.d==1:
			for i in xrange(len(self.x)+1):
				# print 'i = %d, x[i] = '%i,self.x[i]
				if abs(pos-self.x[i])<dx/2.0:
					return i
		elif self.d==2:
			if pos[0]>self.X[0,-1] or pos[0]<self.X[0,0]:
				# print 'this should not happen now'
				pos[0] = np.fmod(pos[0],(self.X[0,-1]-self.X[0,0])+dx)+self.X[0,0]
				pos[0] *= ((self.X[0,-1]-self.X[0,0])+dx) if pos[0]<0 else 1
			for i in xrange(len(self.X)):
				if pos[0]-self.X[i,i]<dx:
					indx[0] = i
					break
			if pos[1]>self.Y[-1,0] or pos[1]<self.Y[0,0]:
				# print 'this should not happen now'
				pos[1] = np.fmod(pos[1],(self.Y[-1,0]-self.Y[0,0])+dy)+self.Y[0,0]
				pos[1] *= ((self.Y[-1,0]-self.Y[0,0])+dx) if pos[1]<0 else 1
				# print pos[1]
			for j in xrange(len(self.Y)):
				if pos[1]-self.Y[j,j]<dy:
					indx[1] = j
					break
			# print indx
			return indx

	def CalculateGradient(self,C):
		"""calculate the concentration gradient in an smart way
		Not so smart yet"""
		grad = np.zeros(np.shape(C))
		if self.d==1:
			for i in xrange(len(C)):
				grad[i] = (C[i]-C[i-1])/(self.x[1]-self.x[0])
		elif self.d==2:
			for i in xrange(len(C)):
				for j in xrange(len(C[i])):
					grad[i,j] = (C[i][j]-C[i-1][j])/((self.X[-1,-1]-self.X[0,0])/len(self.X))
					grad[i,j] = (C[i][j]-C[i][j-1])/((self.Y[-1,-1]-self.Y[0,0])/len(self.Y))
		return grad

	def checkpos(self,r,s,eps = 1e-5):
		"""Implements reflecting boundaries -- Has a bug!"""
		tmp = r+s
		if not self.HasLeftArea(tmp):
			return tmp
		else:
			indx = 0
			b = 1 	#find the relevant boundary
			if self.d==1:
				b = self.x0 if tmp[indx]-self.x0<0 else self.x1
			elif self.d==2:
				if tmp[0]-self.x0<=0:
					indx = 0
					b = self.x0
				elif tmp[0]-self.x1>=0:
					indx = 0
					b = self.x1
				elif tmp[1]-self.y0<=0:
					indx = 1
					b = self.y0
				else:
					indx = 1
					b = self.y1
			f = 0.5
			it = 0
			while f>eps:
				if r[indx]+f*s[indx]-b<=eps:
					r += f*s
					r -= (1-f)*s
					# if it == 0: print 'tmp = ',tmp,' r = ',r,' s = ',s,' f = %g'%f
					return r
				elif r[indx]+f*s[indx]-b <0:
					f += f/2.0
				elif r[indx]+f*s[indx]-b >0:
					f -= f/2.0
				it += 1
			return r+f*s -(1-f)*s



import matplotlib.pyplot as mpl
import matplotlib.animation as animation
from mpl_toolkits.mplot3d import Axes3D 
# area = [[0.3,0.3],[0.4,0.4]]
t =0
def setup_plot():
	mpl.ion()
	fig  = mpl.figure()
	ax = fig.add_subplot(111,projection='3d')
	ax.set_autoscaley_on(False)
	return fig,ax

if __name__ == '__main__':
	if True:
		"1D"
		n = 11
		U = np.zeros(n)
		U[:n/2] = 1
		x = np.linspace(0,1,n)
		area = [[0,1]]
		im = []
		fig  = mpl.figure()
		walk = Walk(area,1.0)
		while t<10:
			print 't = %d'%t
			im.append(mpl.plot(x,U,'b-'))
			U = walk.advance(U)	#[[1,1],[0,0]]
			t+=1
		ani = animation.ArtistAnimation(fig,im,interval=180,blit=True)
		mpl.show()
	if False:
		"2D"
		X,Y = np.meshgrid(np.linspace(0,1,11),np.linspace(0,1,11))
		U = np.zeros((11,11))
		U[:11/2,:11/2] = 1
		area = [[0,0],[1,1]]
		fig,ax = setup_plot()
		walk = Walk(area,1.0)
		wframe = ax.plot_wireframe(X,Y,U)
		mpl.draw()
		while t<10:
			print 't = %d'%t
			ax.collections.remove(wframe)
			U = walk.advance(U)	#[[1,1],[0,0]]
			t+=1
			wframe = ax.plot_wireframe(X,Y,U)
			mpl.draw()

File: code/Walk.cpp
#include "main_walk.h"
using namespace std;

#define IA 16807
#define IM 2147483647
#define AM (1.0/IM)
#define IQ 127773
#define IR 2836
#define MASK 123459876

double ran0(long *idum)
{
   long     k;
   double   ans;

   *idum ^= MASK;
   k = (*idum)/IQ;
   *idum = IA*(*idum - k*IQ) - IR*k;
   if(*idum < 0) *idum += IM;
   ans=AM*(*idum);
   *idum ^= MASK;
   return ans;
}

// long Idum = -1*rand();
long Idum = -1*time(0);
bool debug_walk = false;

Walk::Walk(int dimension)
{
	if(debug_walk){cout<<"Walk::Walk"<<endl;}
	d = dimension;
	steps = 100;
	x0 = 0; x1 = 1;
	y0 = 0; y1 = 1;
	z0 = 0; z1 = 1;
	int D = 1;
	dt = 1.0/steps;
	factor = sqrt(2*D*dt);
	drift = factor/steps;
};

void Walk::SetInitialCondition(int **C, int M, int N){
	/*Takes the initial condition as an array C[m,n] of ints(?) 
	describing how many walkers in each entry. */
	if(debug_walk){cout<<"Walk::SetInitialCondition"<<endl;}
	nwalkers = 0;
	m = M;
	n = N;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			nwalkers += C[i][j];
		}
	}
	walkers = new double*[nwalkers];
	for(int i=0; i<nwalkers;i++){
		walkers[i] = new double[d];
		for(int l=0; l<d;l++){
			walkers[i][l] = 0;
		}
	}
	x = new double[m];
	y = new double[n];
	dx = (x1-x0)/(m-1);
	dy = (n>1)?((y1-y0)/(n-1)):0;
	for(int k=0;k<m;k++){
		x[k] = k*dx;
	}
	for(int k=0;k<n;k++){
		y[k] = k*dy;
	}
	int counter = 0;

	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			for(int l=0; l<C[i][j]; l++){
				PutWalkers(i,j,counter);
				counter ++;
			}
		}
	}
	x0_ = x0 - (dx/2.0);
	x1_ = x1 + (dx/2.0);
	y0_ = y0 - (dx/2.0);
	y1_ = y1 + (dx/2.0);
}

void Walk::ResetInitialCondition(int **C){
	if(debug_walk){cout<<"Walk:ResetInitialCondition"<<endl;}
	ResetWalkers();
	nwalkers = 0;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			nwalkers += C[i][j];
		}
	}
	int counter = 0;
	for(int i=0; i<m;i++){
		for(int j=0; j<n;j++){
			for(int l=0; l<C[i][j]; l++){
				PutWalkers(i,j,counter);
				counter ++;
			}
		}
	}
}
bool Walk::HasLeftArea(double *pos){
	if(debug_walk){cout<<"Walk::HasLeftArea"<<endl;}
	/*pos = [x] in 1d;
	pos = [x,y] in 2d etc*/
	if(d==1){
		if(pos[0]<x0 || pos[0]>x1){
			return true;
		}
	}
	else if(d==2){
		if(pos[0]>x1 || pos[0]<x0){
			return true;
		}
		if(pos[1]<y0 || pos[1]>y1){
			return true;
		}
	}
	return false;
}

void Walk::advance(int **C){
	if(debug_walk){cout<<"Walk::advance"<<endl;}
	double *newPos, **s;
	newPos = new double[d];			//delete at the end!
	s = new double*[steps];			//delete at the end!
	int *index = new int[d];		//delete at the end!
	for(int k=0; k<steps; k++){
		s[k] = new double[d];
		for(int l=0; l<d; l++){
			s[k][l] = 0;
		}
	}
	for(int i=0; i<nwalkers; i++){
		/*For every walker: */
		for(int k=0; k<steps; k++){
			/*Fill array s with steps*d random numbers*/
			for(int l=0; l<d; l++){
				s[k][l] = factor*(0.5-ran0(&Idum));
			}
		}
		for(int j=0;j<steps;j++){
			/*Advance n steps*/
			newPos = Step(walkers[i],s[j]);
			walkers[i] = checkpos(newPos,s[j]);
		}
	}

	if(d==1){
		for(int i=0;i<nwalkers;i++){
			index = FindPosition(walkers[i]);
			C[index[0]] += 1;
		}
	}
	else if(d==2){
		for(int i=0; i<nwalkers; i++){
			index = FindPosition(walkers[i]);
			C[index[0]][index[1]] += 1;
		}
	}
}

double *Walk::Step(double *r,double *s){
	/*Possibility of changing the algorithm*/
	if(debug_walk){cout<<"Walk::Step"<<endl;}
	if(false){
		for(int i=0; i<d; i++){
			r[i] += s[i];
		}
	}
	else if(false){
		for(int i=0; i<d; i++){
			r[i] += dt*dt*s[i];		//Need velocity as well
		}
	}
	else if(true){
		if(d==1){
			r[0] = (s[0]>0) ?(r[0]+factor):(r[0]-factor);
		}
		else if(d==2){
			if(s[0]>s[1]){
				r[0] = (s[0]>0) ?(r[0]+factor):(r[0]-factor);
			}
			else{
				r[1] = (s[1]>0) ?(r[1]+factor):(r[1]-factor);
			}
		}
	}
	r[0] += drift;		//slight drift in x direction
	return r;
}

int Walk::InitializeTimestep(int **C){
	return 0;
}

void Walk::PutWalkers(int i, int j, int counter){
	if(debug_walk){cout<<"Walk::PutWalkers"<<endl;}
	for(int k=0; k<d; k++){
		if(k==0){
			/*Should this be something with the steplength? - cange factor in that case*/
			walkers[counter][k] = x[i]+factor*(0.5-ran0(&Idum));
		}
		else if(k==1){
			walkers[counter][k] = y[j]+factor*(0.5-ran0(&Idum));
		}
	}
}

double **Walk::ReturnBoundary(){
	//should have a better name
	double **tmp;
	return tmp;
}

int *Walk::FindPosition(double *pos){
	/*Maps the walkers position to its index*/
	int *indx;
	if(d==1){
		indx = new int[1];
		for(int i=0; i<m; i++){
			if(fabs(pos[0]-x[i])<dx/2.0){
				indx[0] = i;
				return indx;
			}
		}
	}
	else if(d==2){
		indx = new int[2];
		indx[0] = indx[1] = -1;
		for(int i=0; i<m; i++){
			if(fabs(pos[0]-x[i])<dx/2.0){
				indx[0] = i;
				break;
			}
		}
		for(int j=0; j<n; j++){
			if(fabs(pos[1]-y[j])<dy/2.0){
				indx[1] = j;
				break;
			}
		}
		if(indx[0]==-1||indx[1]==-1){
			cout<<"panic!!"<<endl;
			exit(1);
		}
	}
	return indx;
}
double **Walk::CalculateGradient(){
	double **tmp;
	return tmp;
}
double *Walk::checkpos(double *r,double *s){
	/*Implements reflecting boundaries -- Need to adjust the boundaries by dx/2 so each thing has 
	as much space*/
	// tmp = r+s 	/*This is r in this case*/
	if(not HasLeftArea(r)){
		return r;
	}
	else{
		if(d ==1){
			if (r[0]<x0_){
				r[0] = x0_ - (r[0]-x0_);
				}
			else if(r[0]>x1_){
				r[0] = x1_ - (r[0]-x1_);
				}
			}
		else if(d == 2){
			if(r[0]<x0_){
				r[0] = x0_ - (r[0]-x0_);
				}
			else if( r[0]>x1_){
				r[0] = x1_ - (r[0]-x1_);
				}
			if (r[1]<y0_){
				r[1] = y0_ - (r[1]-y0_);
				}
			else if (r[1]>y1_){
				r[1] = y1_ - (r[1]-y1_);
				}
			}
		return r;
	}
}



// int len(double*){
// 	/*Length function in python*/
// 	double*::iterator it;
// 	while(it != null){
// 		it++;
// 	}
// }
#undef IA
#undef IM
#undef AM
#undef IQ
#undef IR
#undef MASK
#define MBIG 1000000000
#define MSEED 161803398
#define MZ 0
#define FAC (1.0/MBIG)